File: src/main/scala/gameover/fwk/libgdx/gfx/ShaderMemory.scala
package gameover.fwk.libgdx.gfx

import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShaderProgram

import scala.collection.mutable

object ShaderMemory {

  private val shadersPerSpriteBatches = new mutable.HashMap[SpriteBatch, List[ShaderProgram]]

  private def findShaders(batch: SpriteBatch): List[ShaderProgram] = {
    val entries: Option[List[ShaderProgram]] = shadersPerSpriteBatches.get(batch)
    entries match {
      case Some(sp) => sp
      case None => Nil
    }
  }

  def switchShader(batch: SpriteBatch, newShader: ShaderProgram): ShaderProgram = {
    var shaders: List[ShaderProgram] = findShaders(batch)
    var old: ShaderProgram = null
    if (shaders.nonEmpty) {
      old = shaders.head
      shaders = shaders.tail
    }
    shaders = newShader :: shaders
    shadersPerSpriteBatches.put(batch, shaders)
    batch.setShader(newShader)
    old
  }

  def setShader(batch: SpriteBatch, newShader: ShaderProgram) {
    val shaders: List[ShaderProgram] = findShaders(batch)
    shadersPerSpriteBatches.put(batch, newShader :: shaders)
    batch.setShader(newShader)
  }

  def removeCurrentShader(batch: SpriteBatch) {
    var shaders: List[ShaderProgram] = findShaders(batch)
    if (shaders.nonEmpty) {
      shaders = shaders.tail
      shadersPerSpriteBatches.put(batch, shaders)
      if (shaders.nonEmpty) batch.setShader(shaders.head)
    }
  }

  def clearShader(batch: SpriteBatch) {
    shadersPerSpriteBatches.put(batch, Nil)
    batch.setShader(null)
  }
}

File: src/main/kotlin/gameover/fwk/libgdx/gfx/ShaderMemory.kt
package gameover.fwk.libgdx.gfx

import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.glutils.ShaderProgram
import gameover.fwk.collection.head
import gameover.fwk.collection.tail

object ShaderMemory {
    private val shadersPerSpriteBatches = HashMap<SpriteBatch, List<ShaderProgram>>()

    private fun findShaders(batch: SpriteBatch): List<ShaderProgram>? {
        return shadersPerSpriteBatches[batch]
    }

    fun switchShader(batch: SpriteBatch, newShader: ShaderProgram): ShaderProgram? {
        var shaders = findShaders(batch)
        var old: ShaderProgram? = null
        if (shaders != null && shaders.isNotEmpty()) {
            old = shaders.head
            shaders = shaders.tail
        }
        shaders = gameover.fwk.collection.listAppend(newShader, shaders)
        shadersPerSpriteBatches[batch] = shaders
        batch.shader = newShader
        return old
    }

    fun setShader(batch: SpriteBatch, newShader: ShaderProgram) {
        val shaders = findShaders(batch)
        shadersPerSpriteBatches[batch] = gameover.fwk.collection.listAppend(newShader, shaders)
        batch.shader = newShader
    }

    fun removeCurrentShader(batch: SpriteBatch) {
        var shaders = findShaders(batch)
        if (shaders != null && shaders.isNotEmpty()) {
            shaders = shaders.tail
            shadersPerSpriteBatches[batch] = shaders
            if (shaders.isNotEmpty())
                batch.shader = shaders.head
        }
    }

    fun clearShader(batch: SpriteBatch) : List<ShaderProgram>? {
        val shaders = shadersPerSpriteBatches[batch]
        shadersPerSpriteBatches.remove(batch)
        batch.shader = null
        return shaders
    }
}